lib/tree.rb

```diff
@@ -1,64 +1,162 @@
+    return curr_node
+  end
+  
+  # Time Complexity: Average O(log n), worst-case O(n)
+  # Space Complexity: 1
   def find(key)
-    raise NotImplementedError
+    find_helper(@root, key)
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def find_helper(curr_node, key)
+    return if curr_node.nil?
+    if key == curr_node.key
+      return curr_node.value
+    elsif key < curr_node.key
+      find_helper(curr_node.left, key)
+    else 
+      find_helper(curr_node.right, key)
+    end
+  end 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def inorder_helper(curr_node, list)
+    return list if curr_node.nil?
+    inorder_helper(curr_node.left, list)
+    list << {key: curr_node.key, value: curr_node.value}
+    inorder_helper(curr_node.right, list)
+    return list
+  end 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def preorder_helper(curr_node, list)
+    return list if curr_node.nil?
+    list << {key: curr_node.key, value: curr_node.value}
+    preorder_helper(curr_node.left, list)
+    preorder_helper(curr_node.right, list)
+  end 
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
-
-  # Time Complexity: 
-  # Space Complexity: 
+  
+  def postorder_helper(curr_node, list)
+    return list if curr_node.nil?
+    postorder_helper(curr_node.left, list)
+    postorder_helper(curr_node.right, list)
+    list << {key: curr_node.key, value: curr_node.value}
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)...?
   def height
-    raise NotImplementedError
+    return 0 if @root.nil?
+    return [height_helper(@root.left), height_helper(@root.right)].max + 1
   end
-
+  
+  def height_helper(node)
+    return 0 if node.nil?
+    return 1 + [height_helper(node.left), height_helper(node.right)].max
+  end 
+  
   # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def bfs
-    raise NotImplementedError
+    return [] if @root.nil?
+    queue = []
+    list = []
+    queue.push(@root)
+    while queue.length > 0
+      curr_node = queue[0]
+      list << {key: curr_node.key, value: curr_node.value}
+      queue.push(curr_node.left) if curr_node.left
+      queue.push(curr_node.right) if curr_node.right
+      queue.shift
+    end
+    return list 
   end
+  
+  # Time Complexity: Average O(log n), worst case O(n)
+  # Space Complexity: O(1)
+  # Not happy with the time efficiency of this solution, but I couldn't find a way to merge search and assigning more effectively than just calling find from within delete.
 
+  def delete(key)
```

Comment:
This isn't _quite_ working, but it's close.

Also remember you're doing recursion, so you have some space complexity from the call stack.